test/container_app_test.py

```diff
@@ -65,6 +67,41 @@ async def test_container_snapshot_restore(container_client, tmpdir, servicer):
         assert old_client.credentials == ("ta-i-am-restored", "ts-i-am-restored")
 
 
+@pytest.mark.asyncio
+async def test_container_snapshot_restore_heartbeats(tmpdir, servicer):
+    client = _Client(servicer.container_addr, api_pb2.CLIENT_TYPE_CONTAINER, ("ta-123", "task-secret"))
+    async with client as async_client:
+        io_manager = _ContainerIOManager(api_pb2.ContainerArguments(), async_client)
+        restore_path = tmpdir.join("fake-restore-state.json")
+        restore_path.write_text(
+            json.dumps(
+                {
+                    "task_id": "ta-i-am-restored",
+                    "task_secret": "ts-i-am-restored",
+                    "function_id": "fu-i-am-restored",
+                }
+            ),
+            encoding="utf-8",
+        )
+
+        # Ensure that heartbeats do not run before the snapshot
+        # Ensure that heartbeats do run after the snapshot
+        async with io_manager.heartbeats(True):
+            with mock.patch.dict(
+                os.environ,
+                {"MODAL_RESTORE_STATE_PATH": str(restore_path), "MODAL_SERVER_URL": servicer.container_addr},
+            ):
+                with mock.patch("modal.runner.HEARTBEAT_INTERVAL", 1):
+                    assert not list(
+                        filter(lambda req: isinstance(req, api_pb2.ContainerHeartbeatRequest), servicer.requests)
+                    )
+                    await io_manager.memory_snapshot()
+                    await asyncio.sleep(1)
+                    assert list(
+                        filter(lambda req: isinstance(req, api_pb2.ContainerHeartbeatRequest), servicer.requests)
```

Comment:
Put a sleep here to ensure a heartbeat actually gets sent (heartbeat interval is 1s) @thundergolfer